File: untrue.hs
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as C
import qualified Data.List as L
import Data.Word

import Data.Bits
import Data.Char
import Data.Int
import Data.Maybe

main = do
  contents <- B.getContents
  B.putStr $ generate $ tokenize contents

data Primitive = Ret | Add | Sub | Mul | Div | Neg | Eq | Gt | And | Or | Not |
                 Store | Fetch | Call | Dup | Drop | Swap | Rot | Pick | If |
                 While | Puts | Puti | Putc | Getc
  deriving (Show, Eq)

-- Lookup a primitive from its surface form (one character)
lookup_primitive :: Char -> Maybe Primitive
lookup_primitive c =
  lookup c [('+', Add), ('-', Sub), ('*', Mul), ('/', Div), ('_', Neg),
            ('=', Eq), ('>', Gt), ('&', And), ('|', Or), ('~', Not),
            (':', Store), (';', Fetch), ('!', Call), ('$', Dup), ('%', Drop),
            ('\\', Swap), ('@', Rot), ('`', Pick), ('?', If), ('#', While),
            ('.', Puti), (',', Putc), ('^', Getc)]

opcodes = [(Ret, 0x80), (Add, 0x81), (Sub, 0x82), (Mul, 0x83), (Div, 0x84),
           (Neg, 0x85), (Eq, 0x86), (Gt, 0x87), (And, 0x88), (Or, 0x89),
           (Not, 0x8a), (Store, 0x8b), (Fetch, 0x8c), (Call, 0x8d), (Dup, 0x8e),
           (Drop, 0x8f), (Swap, 0x90), (Rot, 0x91), (Pick, 0x92), (If, 0x93),
           (While, 0x94), (Puts, 0x95), (Puti, 0x96), (Putc, 0x97),
           (Getc, 0x98)]

opcode :: Primitive -> Char
opcode p = chr $ fromJust $ lookup p opcodes

data Token = Function Primitive | LiteralString C.ByteString |
             LiteralNumber Int | Global Char | Lambda [Token] (Maybe Token)
  deriving Show


data TokenizerState = Chunk | StringToken C.ByteString | NumberToken Int |
                      Comment | Quote

-- Tokenize the string into a Lambda token
tokenize :: B.ByteString -> Token
tokenize s =
  let globals = [LiteralNumber 0 | x <- ['a' .. 'z']]
      start = 6 + 4 * length globals
  in tokenize' Chunk
       (Lambda ([LiteralNumber start, Function Call, Function Ret] ++ globals)
       Nothing) (C.unpack s)

tokenize' :: TokenizerState -> Token -> String -> Token
tokenize' Chunk token@(Lambda _ Nothing) "" = token
tokenize' Chunk p ('[':xs) = tokenize' Chunk (Lambda [] (Just p)) xs
tokenize' Chunk (Lambda ts (Just (Lambda ts' p))) (']':xs) =
  tokenize' Chunk (Lambda (ts' ++ [Lambda ts Nothing]) p) xs
tokenize' Chunk p ('{':xs) = tokenize' Comment p xs
tokenize' Comment p ('}':xs) = tokenize' Chunk p xs
tokenize' Comment p (_:xs) = tokenize' Comment p xs
tokenize' Chunk p ('"':xs) = tokenize' (StringToken C.empty) p xs
tokenize' (StringToken s) (Lambda ts p) ('"':xs) =
  tokenize' Chunk (Lambda (ts ++ [LiteralString s]) p) xs
tokenize' (StringToken s) t (x:xs) = tokenize' (StringToken (C.snoc s x)) t xs
tokenize' Chunk t ('\'':xs) = tokenize' Quote t xs
tokenize' Quote (Lambda ts p) (x:xs) =
  tokenize' Chunk (Lambda (ts ++ [LiteralNumber $ ord x]) p) xs
tokenize' Chunk (Lambda ts p) (x:xs)
  | x >= '0' && x <= '9' =
      tokenize' (NumberToken $ (ord x - ord '0')) (Lambda ts p) xs
  | x >= 'a' && x <= 'z' =
      tokenize' Chunk (Lambda (ts ++ [Global x]) p) xs
  | otherwise = case lookup_primitive x of
      (Just op) -> tokenize' Chunk (Lambda (ts ++ [Function op]) p) xs
      Nothing -> tokenize' Chunk (Lambda ts p) xs
tokenize' (NumberToken n) (Lambda ts p) "" = Lambda (ts ++ [LiteralNumber n]) p
tokenize' (NumberToken n) (Lambda ts p) (x:xs)
  | x >= '0' && x <= '9' =
      tokenize' (NumberToken $ 10 * n + (ord x - ord '0')) (Lambda ts p) xs
  | otherwise = tokenize' Chunk (Lambda (ts ++ [LiteralNumber n]) p) (x:xs)

-- Generate code for the body of the top lambda expression
generate :: Token -> C.ByteString
generate (Lambda xs _) = generate' 0 0 xs

generate' :: Int -> Int -> [Token] -> C.ByteString
generate' _ _ [] = generate'' (Function Ret)
generate' m n ((Function op):xs) =
  C.append (generate'' $ Function op) (generate' m (n + 1) xs)
generate' m n ((LiteralString s):xs) = let s' = generate'' $ LiteralString s in
  C.append s' (generate' m (n + C.length s') xs)
generate' m n ((LiteralNumber n'):xs) =
  let n'' = generate'' $ LiteralNumber n'
  in C.append n'' (generate' m (n + C.length n'') xs)
generate' m n ((Global c):xs) =
  let c' = generate'' $ LiteralNumber $ 4 * (ord c - ord 'a') + 6
  in C.append c' (generate' m (m + C.length c') xs)
generate' m n ((Lambda ys _):xs) =
  let xs' = generate' m (n + 4) xs
      m' = m + n + 4 + C.length xs'
      ys' = generate' m' m' ys
  in C.append (generate'' $ LiteralNumber m') $ C.append xs' ys'

generate'' :: Token -> C.ByteString
generate'' (Function op) = C.singleton $ opcode op
generate'' (LiteralString s) =
  C.append (generate'' $ Function Puts) $
    C.append (generate'' $ LiteralNumber $ C.length s) s
generate'' (LiteralNumber n) =
  C.pack $ map chr $
    (shift (n .&. 0x7f000000) (-24)) : (shift (n .&. 0xff0000) (-16)) :
    (shift (n .&. 0xff00) (-8)) : (n .&. 0xff) : []

File: untrue.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>

#define STRING_SIZE 256

// Primitives
enum { RET = 0x80, ADD, SUB, MUL, DIV, NEG, EQ, GT, AND, OR, NOT, STORE, FETCH,
  CALL, DUP, DROP, SWAP, ROT, PICK, IF, WHILE, PUTS, PUTI, PUTC, GETC, NOP };

// Tokenizer state types
enum { chunk, string_token, number_token, comment, quote };

// Token types
enum { primitive, literal_string, literal_number, global, lambda };

// Tokens
typedef struct token {
  size_t type;
  union {
    size_t n;                   // numerical value: primitive, integer, address
    char *string;               // for literal strings
    struct token_list *body;    // for lambda
  } value;
  struct token *next;
} token;

typedef struct token_list {
  token *first;
  token *last;
  struct token_list *parent;
} token_list;

size_t lookup_primitive(int);
token_list *make_token_list(token_list *);
token_list *append(token_list *, token *);
token *make_lambda(token_list *);
token_list *tokenize(FILE *);

size_t lookup_primitive(int c) {
  switch (c) {
    case '+': return ADD;
    case '-': return SUB;
    case '*': return MUL;
    case '/': return DIV;
    case '_': return NEG;
    case '=': return EQ;
    case '>': return GT;
    case '&': return AND;
    case '|': return OR;
    case '~': return NOT;
    case ':': return STORE;
    case ';': return FETCH;
    case '!': return CALL;
    case '$': return DUP;
    case '\\': return DROP;
    case '%': return SWAP;
    case '@': return ROT;
    case '`': return PICK;
    case '?': return IF;
    case '#': return WHILE;
    case '.': return PUTI;
    case ',': return PUTC;
    case '^': return GETC;
  }
  return 0;
}

token_list *make_token_list(token_list *parent) {
  token_list *l = (token_list *)malloc(sizeof(token_list));
  if (l) {
    l->first = NULL;
    l->last = NULL;
    l->parent = parent;
  }
  return l;
}

token_list *append(token_list *l, token *t) {
  if (l->last) {
    l->last->next = t;
  } else {
    l->first = t;
  }
  l->last = t;
  return l;
}

token *make_lambda(token_list *parent) {
  token *t = (token *)malloc(sizeof(token));
  if (t) {
    t->value.body = make_token_list(parent);
    if (t->value.body) {
      t->type = lambda;
    } else {
      t = NULL;
    }
  }
  return t;
}

token_list *tokenize(FILE *f) {
  int state = chunk;
  token_list *tokens = make_token_list(NULL);
  token *t;
  char *string = NULL;
  size_t n = 0;
  for (int c = fgetc(f); !feof(f); c = fgetc(f)) {
next_char: switch (state) {
      case chunk:
        if (c == '[') {
          t = make_lambda(tokens);
          append(tokens, t);
          tokens = t->value.body;
        } else if (c == ']') {
          tokens = tokens->parent;
        } else if (c == '{') {
          state = comment;
          n = 1;
        } else if (c == '"') {
          state = string_token;
          string = (char *)malloc(STRING_SIZE);
          n = 0;
        } else if (c == '\'') {
          state = quote;
        } else if (c >= '0' && c <= '9') {
          state = number_token;
          n = (unsigned char) c - '0';
        } else if (c >= 'a' && c <= 'z') {
          t = (token *)malloc(sizeof(token));
          t->type = literal_number;
          t->value.n = 4 * ((unsigned char) c - 'a' + 2);
          append(tokens, t);
        } else {
          n = lookup_primitive(c);
          if (n > RET && n < NOP) {
            t = (token *)malloc(sizeof(token));
            t->type = primitive;
            t->value.n = n;
            append(tokens, t);
          }
        }
        break;
      case string_token:
        if (c == '"') {
          string[n++] = '\0';
          t = (token *)malloc(sizeof(token));
          t->type = literal_string;
          t->value.string = (char *)realloc(string, n);
          append(tokens, t);
          state = chunk;
        } else {
          string[n++] = (char) c;
          if (n % STRING_SIZE == 0) {
            string = (char *)realloc(string, n + STRING_SIZE);
          }
        }
        break;
      case number_token:
        if (c >= '0' && c <= '9') {
          n = 10 * n + ((unsigned char) c - '0');
        } else {
          t = (token *)malloc(sizeof(token));
          t->type = literal_number;
          t->value.n = n;
          append(tokens, t);
          state = chunk;
          goto next_char;
        }
        break;
      case comment:
        if (c == '{') {
          ++n;
        } else if (c == '}') {
          if (--n == 0) {
            state = chunk;
          }
        }
        break;
      case quote:
        t = (token *)malloc(sizeof(token));
        t->type = literal_number;
        t->value.n = (unsigned char) c;
        append(tokens, t);
        state = chunk;
    }
  }
  return tokens;
}

int main(int argc, char *argv[]) {
  if (argc < 1) {
    return 1;
  }
  FILE *f = strcmp(argv[1], "-") == 0 ? stdin : fopen(argv[1], "r");
  if (!f) {
    return 1;
  }
  static struct termios oldt, newt;
  tcgetattr(STDIN_FILENO, &oldt);
  newt = oldt;
  if (strcmp(argv[1], "-") != 0) {
    newt.c_lflag &= (unsigned long) ~(ICANON);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  }
  (void) tokenize(f);
  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
  return 0;
}

